### PR TITLE
Add capability to specify auth from env var or stdin

### PR DESCRIPTION
**Pull request by aeisenberg**
This commit adds two new ways of specifying GitHub auth:

1. from the GITHUB_TOKEN environment variable
2. from standard input

This commit does not include any documentation changes and the
descriptions of new command line options will need to be tweaked.
I will work on that later.

### Merge / deployment checklist

- [x] Confirm this change is backwards compatible with existing workflows.
- [x] Confirm the [readme](https://github.com/github/codeql-action/blob/master/README.md) has been updated if necessary.

### CODE REVIEW

File: src/util.ts [hunk 1]
```diff
@@ -301,3 +302,60 @@ export function apiVersionInRange(
...
+  githubAuth: string | undefined,
+  fromStdIn: boolean | undefined,
+  readable = process.stdin as Readable
+): Promise<string> {
+  if (githubAuth && fromStdIn) {
+    throw new Error("Cannot specify both '--github-auth' and '--with-token'.");
+  }
+
+  if (githubAuth) {
+    return githubAuth;
```
Comment (adityasharad) [inline src/util.ts:333]: Can we log a deprecation warning here? As a reference, Docker uses `WARNING! Using --password via the CLI is insecure. Use --password-stdin.` (We could also adopt that naming convention: `--github-auth-stdin`).

File: src/util.ts [hunk 2]
```diff
@@ -301,3 +302,60 @@ export function apiVersionInRange(
...
+ *
+ * @return a promise resolving to the auth token.
+ */
+export async function getGitHubAuth(
+  githubAuth: string | undefined,
+  fromStdIn: boolean | undefined,
+  readable = process.stdin as Readable
+): Promise<string> {
+  if (githubAuth && fromStdIn) {
+    throw new Error("Cannot specify both '--github-auth' and '--with-token'.");
```
Comment (adityasharad) [inline src/util.ts:329]: Recommend that the user uses the stdin option.

File: src/util.ts [hunk 3]
```diff
@@ -301,3 +302,60 @@ export function apiVersionInRange(
...
+  }
+
+  if (fromStdIn) {
+    return new Promise((resolve, reject) => {
+      let token = "";
+      readable.on("data", (data) => {
+        token += data.toString("utf8");
+      });
+      readable.on("end", () => {
+        token = token.split(/\s+/)[0].trim();
```
Comment (adityasharad) [inline src/util.ts:343]: Make sure you don't crash here if it's empty.

File: src/util.ts [hunk 4]
```diff
@@ -301,3 +302,60 @@ export function apiVersionInRange(
...
+        reject(err);
+      });
+    });
+  }
+
+  if (process.env.GITHUB_TOKEN) {
+    return process.env.GITHUB_TOKEN;
+  }
+
+  throw new Error("No GitHub authentication token is specified.");
```
Comment (adityasharad) [inline src/util.ts:360]:
```suggestion
  throw new Error("No GitHub authentication token was specified. Please provide a token via the GITHUB_TOKEN environment variable, or by adding the `--<flagname>` flag and passing the token on standard input.");
```

File: src/runner.ts [hunk 5]
```diff
@@ -442,9 +451,13 @@ program
   )
   .requiredOption("--ref <ref>", "Name of ref that was analyzed. (Required)")
   .requiredOption("--github-url <url>", "URL of GitHub instance. (Required)")
-  .requiredOption(
+  .option(
     "--github-auth <auth>",
-    "GitHub Apps token or personal access token. (Required)"
+    "GitHub Apps token or personal access token. This option is insecure...(WIP come up with a better description)"
```
Comment (adityasharad) [inline src/runner.ts:456]: Suggest something like "Using this option is insecure, and will be deprecated. Please use --<safer option>."

**Review by adityasharad — COMMENTED** [review]
Thanks! Generally looks sensible, suggestions are mainly about user-facing wording.

**Comment by aeisenberg** [comment]
All your suggestions make sense. Let me work on that.

File: src/util.ts (hunk 3 above)
Comment (aeisenberg) [reply src/util.ts:343]: I don't see how there is a chance of crashingl

Comment (adityasharad) [reply src/util.ts:343]: Ah if `token` is empty, split gives a singleton list. Okay nothing to see here :)

Comment (adityasharad) [reply src/util.ts:343]: Ah if `token` is empty, split gives a singleton list. Okay nothing to see here :)

**Comment by aeisenberg** [comment]
I missed an instance of `--github-auth`. It's added now.

**Review by robertbrignull — COMMENTED** [review]
Haven't formally approved yet. I'll let @adityasharad do that once the requirements are nailed down. But just saying that the code here LGTM from a low-level code point of view.

**Comment by aeisenberg** [comment]
Is there any documentation change that needs to occur? Or are the new cli options sufficient?

**Review by adityasharad — APPROVED** [review]
Looks good. Docs team suggested we file an issue internally to update the Code Scanning help.

**Comment by JLLeitschuh** [comment]
Is there a plan to issue a CVE number for this?

**Pull request merged**